tsl/src/continuous_aggs/invalidation_record.c

```diff
@@ -0,0 +1,120 @@
+
+TS_FUNCTION_INFO_V1(ts_invalidation_read_record);
+
+/*
+ * Helper macros always use the current version when encoding the entry.
+ */
+#define _field(v) CppConcat(ver, v)
+#define _ver _field(INVALIDATION_ENTRY_CURRENT_VERSION)
+
+/*
+ * Read an encoded invalidation record coming from the plugin.
+ *
+ * Records from the plugin is coming in logical replication format, so decode
+ * it and produce a single composite value in the same format as the
+ * invalidation table.
+ */
+Datum
+ts_invalidation_read_record(PG_FUNCTION_ARGS)
+{
+	bytea *raw_record = PG_GETARG_BYTEA_P(0);
+
+	TupleDesc tupdesc;
+	if (get_call_result_type(fcinfo, NULL, &tupdesc) != TYPEFUNC_COMPOSITE)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("function returning record called in context "
+						"that cannot accept type record")));
+	tupdesc = BlessTupleDesc(tupdesc);
+
+	StringInfoData info;
+	initReadOnlyStringInfo(&info, VARDATA_ANY(raw_record), VARSIZE_ANY_EXHDR(raw_record));
+
+	InvalidationEntry entry;
+	ts_invalidation_record_decode(&info, &entry);
+
+	Datum values[] = {
+		[0] = ObjectIdGetDatum(entry._ver.hypertable_relid),
+		[1] = Int64GetDatum(entry._ver.lowest_modified),
+		[2] = Int64GetDatum(entry._ver.highest_modified),
+	};
+	bool nulls[3] = { 0 };
+
+	Assert(tupdesc->natts == sizeof(values) / sizeof(*values));
+
+	HeapTuple htup = heap_form_tuple(tupdesc, values, nulls);
+
+	PG_RETURN_DATUM(HeapTupleGetDatum(htup));
+}
+
+/*
+ * Decode invalidation record in byte array format.
+ *
+ * This need to handle all supported version formats, not just the current
+ * one. This function should be used on the receiving side of the invalidation
+ * plugin and be used to decode records arriving from the plugin.
+ */
+void
+ts_invalidation_record_decode(StringInfo record, InvalidationEntry *entry)
+{
+	/*
+	 * The header is the same for all versions of the invalidation record.
+	 */
+	entry->version = pq_getmsgint32(record);
+
+	switch (entry->version)
+	{
+		case 1:
+			entry->ver1.hypertable_relid = pq_getmsgint32(record);
+			entry->ver1.lowest_modified = pq_getmsgint64(record);
+			entry->ver1.highest_modified = pq_getmsgint64(record);
+			break;
+		default:
+			ereport(ERROR,
+					errcode(ERRCODE_PROTOCOL_VIOLATION),
+					errmsg("unrecognized invalidation record version %d", entry->version),
+					errdetail(
+						"Invalidation record from invalidation plugin is using an unrecognized "
+						"version, suggesting that it is newer than the extension."),
+					errhint("Make sure that invalidation plugin is not from a newer build than the "
+							"extension."));
+			break;
+	}
+}
+
+/*
+ * Encode an invalidation tuple in byte array format.
+ */
+void
+ts_invalidation_record_encode(InvalidationEntry *entry, StringInfo record)
+{
+	/*
+	 * Header is always the same for all versions. We deliberately set the
+	 * version here to avoid bugs in calling this function.
+	 */
+	entry->version = INVALIDATION_ENTRY_CURRENT_VERSION;
```

Comment:
Not sure why we need to write this version to the struct. I don't think it is ever used anywhere....

Why not just serialize the version directly to the output?

Reply:
I want to keep the same structure available on both the reader and the writer side since that makes debugging and resolving issues a lot easier, especially on the reading side. There are many ways to skin a cat though, but I picked this one.